go/database/mpt/proof.go

```diff
@@ -0,0 +1,115 @@
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package mpt
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"github.com/Fantom-foundation/Carmen/go/common"
+)
+
+//go:generate mockgen -source proof.go -destination proof_mocks.go -package witness
+
+// rlpEncodedNode is an RLP encoded MPT node.
+type rlpEncodedNode []byte
+
+// rlpEncodedNodeEquals returns true if the two RLP encoded nodes are equal.
+func rlpEncodedNodeEquals(a, b rlpEncodedNode) bool {
+	return bytes.Equal(a, b)
+}
+
+// proofDb is a database of RLP encoded MPT nodes and their hashes that represent witness proofs.
+type proofDb map[common.Hash]rlpEncodedNode
+
+// WitnessProof represents a witness proof.
+// It contains a database of MPT nodes and their hashes.
+// The proof may be verified by iterating the proof database from the
+// root hash to a terminal entry according to a Path
+// representing either an account address or a storage key.
+// If the proof contains an account or a storage slot for the input address or key,
+// the proof is representing this account or storage.
+type WitnessProof proofDb
+
+// CreateWitnessProof creates a witness proof for the input account address
+// and possibly storage slots of the same account under the input storage keys.
+// If the proof cannot be created because either the address or a key does not exist,
+// it returns an error.
+// It may return an error also when it occurs in the underlying database.
+func CreateWitnessProof(nodeSource NodeSource, root *NodeReference, address common.Address, keys ...common.Key) (WitnessProof, error) {
+	proof := WitnessProof{}
+	accountVisitor := &proofExtractionVisitor{
+		nodeSource: nodeSource,
+		proof:      proof,
+	}
+
+	var innerError error
+
+	found, err := VisitPathToAccount(nodeSource, root, address, MakeVisitor(func(node Node, info NodeInfo) VisitResponse {
+		if res := accountVisitor.Visit(node, info); res == VisitResponseAbort {
+			return VisitResponseAbort
+		}
+		// if account reached, prove storage keys and terminate.
+		switch account := node.(type) {
+		case *AccountNode:
+			for _, key := range keys {
+				storageVisitor := &proofExtractionVisitor{
+					nodeSource: nodeSource,
+					proof:      proof,
+				}
+				found, err := VisitPathToStorage(nodeSource, &account.storage, key, storageVisitor)
+				if err != nil || storageVisitor.err != nil {
+					innerError = errors.Join(innerError, storageVisitor.err, err)
+					return VisitResponseAbort
+				}
+				if !found {
+					innerError = errors.Join(innerError, fmt.Errorf("storage key %s not found", key))
+				}
+			}
+			return VisitResponseAbort
+		}
+
+		return VisitResponseContinue
+	}))
+
+	if !found {
+		err = errors.Join(err, fmt.Errorf("account %s not found", address))
+	}
```

Comment:
Not finding the account is not an error. It means the account is empty, which is fine. A proof for an empty account is valid and useful.

Reply:
I see, ok I will basically ignore not-found